Cantera does not check charge balance (!)

**Problem description**

Cantera does not check charge balance, and this is problematic for any mechanism containing ions and electrons.

**Steps to reproduce**

In _mechanism_with_problem.yaml_:
```
units: {length: cm, time: s, quantity: mol, activation-energy: kJ/mol}

phases:
- name: isotropic-electron-energy-plasma
  thermo: plasma
  elements: [O, E]
  species: [e-, O, O+]
  kinetics: gas
  state: {T: 300.0, P: 1 atm}
  electron-energy-distribution:  # https://cantera.org/dev/yaml/phases.html#plasma
    type: isotropic  # isotropic or discretized
    shape-factor: 1.0   # 1.0 for Maxwellian, 2 for Druyvesteyn
    energy-levels: [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0]  # eV
    mean-electron-energy: 2.0 eV # By default, T_e = T_gas, and e=3/2 T_e
- name: gas
  thermo: ideal-gas
  elements: [O, E]
  species: [e-, O, O+]
  kinetics: gas
  state: {T: 300.0, P: 1 atm}

species:
- name: e-
  composition: {E: 1}
  thermo:
    model: NASA9
    temperature-ranges: [298.15, 1000.0, 6000.0, 2.0e+04]
    data:
    - [0.0, 0.0, 2.5, 0.0, 0.0, 0.0, 0.0, -745.375, -11.72081224]
    - [0.0, 0.0, 2.5, 0.0, 0.0, 0.0, 0.0, -745.375, -11.72081224]
    - [0.0, 0.0, 2.5, 0.0, 0.0, 0.0, 0.0, -745.375, -11.72081224]
    note: Ref-Species. Chase,1998 3/82. [g12/98]
- name: O
  composition: {O: 1}
  thermo:
    model: NASA9
    temperature-ranges: [200.0, 1000.0, 6000.0, 2.0e+04]
    data:
    - [-7953.6113, 160.7177787, 1.966226438, 1.01367031e-03, -1.110415423e-06, 6.5175075e-10, -1.584779251e-13, 2.840362437e+04, 8.40424182]
    - [2.619020262e+05, -729.872203, 3.31717727, -4.28133436e-04, 1.036104594e-07, -9.43830433e-12, 2.725038297e-16, 3.39242806e+04, -0.667958535]
    - [1.779004264e+08, -1.082328257e+05, 28.10778365, -2.975232262e-03, 1.854997534e-07, -5.79623154e-12, 7.191720164e-17, 8.89094263e+05, -218.1728151]
    note: D0(O2):Brix,1954. Moore,1976. Gordon,1999. [g 5/97]
- name: O+
  composition: {O: 1, E: -1}
  thermo:
    model: NASA9
    temperature-ranges: [298.15, 1000.0, 6000.0, 2.0e+04]
    data:
    - [0.0, 0.0, 2.5, 0.0, 0.0, 0.0, 0.0, 1.879352842e+05, 4.39337676]
    - [-2.166513208e+05, 666.545615, 1.702064364, 4.71499281e-04, -1.427131823e-07, 2.016595903e-11, -9.107157762e-16, 1.837191966e+05, 10.05690382]
    - [-2.143835383e+08, 1.469518523e+05, -36.8086454, 5.03616454e-03, -3.087873854e-07, 9.18683487e-12, -1.074163268e-16, -9.61420896e+05, 342.619308]
    note: Martin,W.C.,1993. Gordon,1999. [g 8/97]

reactions:
- equation: e- + O+ => O  # Reaction 1: Ok
  type: two-temperature-plasma
  rate-constant: {A: 7.43e+12, b: -0.29, Ea-gas: 0.00e+03, Ea-electron: 0.00e+03}
  umist-url: https://umistdatabase.uk/react/8741
- equation: e- + O+ => O+  # Reaction 2: NOT OK AT ALL
  type: two-temperature-plasma
  rate-constant: {A: 7.43e+12, b: -0.29, Ea-gas: 0.00e+03, Ea-electron: 0.00e+03}
```

In _cantera_test_mechanism_with_problem.py_
```
import cantera as ct

plasma = ct.Solution(
    PATH_TO --> "mechanism_with_problem.yaml",  # Change with the path to the previous mechanism
    name="isotropic-electron-energy-plasma",
)


# List of reactions.
reactions = plasma.reactions()
print(f"Number of reactions: {len(reactions)}")

print(f"Reaction 1: {reactions[0]}")
# >>> Reaction 1: O+ + e- => O

# Reaction 2: PROBLEM HERE --> Cantera does not throw an error, but charges are not balanced.
print(f"Reaction 2: {reactions[1]}")
# Reaction 2: O+ + e- => O+
```


**Behavior**

I would expect Cantera to throw an error when charge balance is not respected.
For instance, reaction 1 is OK and reaction 2 is NOT OK.


**System information**

- Cantera version: 3.1.0
- OS: Windows 11
- Python version: 3.11.10

Thanks for reporting this. Cantera checks reactions for conservation of "elements" where electrons are treated as an element that needs to be conserved. Unfortunately, there seems to be a bug in how this check is handled in some cases where you have negative numbers of an element as part of a species composition, which is how we represent positive ions. You can see that this check does catch some other charge balance errors if you add a reaction like:
```yaml
- equation: e- + O => O
  rate-constant: {A: 7.43e+12, b: -0.29, Ea: 0.00e+03}
```
which produces the error:
```
*******************************************************************************
CanteraError thrown by addReactions:

*******************************************************************************
InputFileError thrown by Reaction::checkBalance:
Error on line 82 of ./balance.yaml:
The following reaction is unbalanced: O + e- => O
  Element    Reactants    Products
  E           1           0

|  Line |
|    77 |   rate-constant: {A: 7.43e+12, b: -0.29, Ea-gas: 0.00e+03, Ea-electron: 0.00e+03}
|    78 |   umist-url: https://umistdatabase.uk/react/8741
|    79 | - equation: e- + O+ => O+  # Reaction 2: NOT OK AT ALL
|    80 |   type: two-temperature-plasma
|    81 |   rate-constant: {A: 7.43e+12, b: -0.29, Ea-gas: 0.00e+03, Ea-electron: 0.00e+03}
>    82 > - equation: e- + O => O
            ^
|    83 |   rate-constant: {A: 7.43e+12, b: -0.29, Ea: 0.00e+03}
|    84 | 
|    85 | 
*******************************************************************************
*******************************************************************************
```

--- a/include/cantera/thermo/HMWSoln.h
+++ b/include/cantera/thermo/HMWSoln.h
@@ -1386,7 +1386,7 @@ private:
      *          dw = C_0 * M_0 (density of water) (kg/m3)
      *             = 1.0E3 at 25C
      */
-    mutable double m_A_Debye;
+    mutable double m_A_Debye = 1.172576;
 
     //! Water standard state calculator
     /*!

--- a/src/kinetics/Reaction.cpp
+++ b/src/kinetics/Reaction.cpp
@@ -624,9 +624,9 @@ void Reaction::checkBalance(const Kinetics& kin) const
     string msg;
     bool ok = true;
     for (const auto& [elem, balance] : balr) {
-        double elemsum = balr[elem] + balp[elem];
+        double scale = std::max(std::abs(balr[elem]), std::abs(balp[elem]));
         double elemdiff = fabs(balp[elem] - balr[elem]);
-        if (elemsum > 0.0 && elemdiff / elemsum > 1e-4) {
+        if (elemdiff > 1e-4 * scale) {
             ok = false;
             msg += fmt::format("  {}           {}           {}\n",
                                elem, balr[elem], balp[elem]);

--- a/test/SConscript
+++ b/test/SConscript
@@ -177,7 +177,7 @@ def addPythonTest(testname, subset):
                 if class_name.startswith("build.python.cantera.test."):
                     class_name = class_name[26:]
                 test_name = "python: {}.{}".format(class_name, test.get('name'))
-                if test.findall('failure'):
+                if test.findall('failure') or test.findall('error'):
                     test_results.failed[test_name] = 1
                     failures += 1
                 else:

--- a/test/data/kineticsfromscratch.yaml
+++ b/test/data/kineticsfromscratch.yaml
@@ -12,7 +12,6 @@ units: {activation-energy: cal/mol}
 phases:
 - name: ohmech
   thermo: ideal-gas
-  elements: [O, H, Ar]
   species:
   - h2o2.yaml/species: [AR, O, H2, H, OH, O2, H2O, H2O2, HO2]
   kinetics: gas
@@ -32,6 +31,14 @@ phases:
     T: 900.0
     coverages: {PT(S): 0.35, H(S): 0.35, H2O(S): .1, OH(S): 0.1, O(S): 0.1}
   site-density: 2.7063e-09
+- name: ions
+  thermo: ideal-gas
+  species:
+  - h2o2.yaml/species: [AR, O, H2, H, OH, O2, H2O, H2O2, HO2]
+  - ch4_ion.yaml/species: [E, O2-, H3O+]
+  kinetics: gas
+  reactions: none
+  state: {T: 300.0, P: 1.01325e+05}
 
 reactions:
 - equation: O + H2 <=> H + OH  # Reaction 1

--- a/test/kinetics/kineticsFromScratch.cpp
+++ b/test/kinetics/kineticsFromScratch.cpp
@@ -567,6 +567,38 @@ TEST_F(KineticsFromScratch, invalid_nonreactant_order)
     ASSERT_EQ((size_t) 0, kin.nReactions());
 }
 
+TEST_F(KineticsFromScratch, unbalanced_neutral)
+{
+    Composition reac = parseCompString("O:1 H2:1");
+    Composition prod = parseCompString("H:1 H2O:1");
+    auto rate = make_shared<ArrheniusRate>(1.0, 0.0, 0.0);
+    auto R = make_shared<Reaction>(reac, prod, rate);
+    ASSERT_THROW(kin.addReaction(R), CanteraError);
+    ASSERT_EQ((size_t) 0, kin.nReactions());
+}
+
+TEST(KineticsFromScratch2, unbalanced_positive_ion)
+{
+    auto soln = newSolution("kineticsfromscratch.yaml", "ions");
+    Composition reac = parseCompString("H3O+:1");
+    Composition prod = parseCompString("H2O:1 H:1");
+    auto rate = make_shared<ArrheniusRate>(1.0, 0.0, 0.0);
+    auto R = make_shared<Reaction>(reac, prod, rate);
+    ASSERT_THROW(soln->kinetics()->addReaction(R), CanteraError);
+    ASSERT_EQ((size_t) 0, soln->kinetics()->nReactions());
+}
+
+TEST(KineticsFromScratch2, unbalanced_negative_ion)
+{
+    auto soln = newSolution("kineticsfromscratch.yaml", "ions");
+    Composition reac = parseCompString("O2-:1");
+    Composition prod = parseCompString("O2:1 E:2");
+    auto rate = make_shared<ArrheniusRate>(1.0, 0.0, 0.0);
+    auto R = make_shared<Reaction>(reac, prod, rate);
+    ASSERT_THROW(soln->kinetics()->addReaction(R), CanteraError);
+    ASSERT_EQ((size_t) 0, soln->kinetics()->nReactions());
+}
+
 class InterfaceKineticsFromScratch : public testing::Test
 {
 public:

